File: client/lib/api/schemas/posts.ts
import { z } from "zod";

export const FORUM_POST_SOURCES = ["p2pfoundation", "bitcointalk"] as const;
export const zForumPostSource = z.enum(FORUM_POST_SOURCES);
export type ForumPostSource = z.infer<typeof zForumPostSource>;

const zForumPost = z.object({
  date: z.coerce.date(),
  subject: z.string(),
  text: z.string(),
  source: zForumPostSource,
  sourceId: z.string(),
  url: z.string(),
  threadId: z.number().int().min(1),
  satoshiId: z.number().int().min(1).nullable(),
  nestedLevel: z.number().int().min(0),
  posterName: z.string(),
  posterUrl: z.string().nullable(),
});
export type ForumPost = z.infer<typeof zForumPost>;

export const zSatoshiForumPost = zForumPost.extend({
  satoshiId: z.number().int().min(1),
});
export type SatoshiForumPost = z.infer<typeof zSatoshiForumPost>;

export const zForumPostIndex = z.array(zSatoshiForumPost);

export const zForumPostDetail = z.object({
  post: zSatoshiForumPost,
  previous: zSatoshiForumPost.nullable(),
  next: zSatoshiForumPost.nullable(),
});
export type ForumPostDetail = z.infer<typeof zForumPostDetail>;

export const zForumThread = z.object({
  id: z.number().int().min(1),
  title: z.string(),
  source: zForumPostSource,
  date: z.coerce.date(),
  url: z.string(),
});
export type ForumThread = z.infer<typeof zForumThread>;

export const zForumThreadIndex = z.array(zForumThread);

export const zForumThreadDetail = z.object({
  thread: zForumThread,
  posts: z.array(zForumPost),
  previous: zForumThread.nullable(),
  next: zForumThread.nullable(),
});

File: server/sni/models/satoshi/posts.py
import datetime
from typing import TYPE_CHECKING

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from sni.database import Base
from sni.models.content import JSONContent

if TYPE_CHECKING:
    from sni.models.satoshi.quotes import Quote


class ForumThread(Base):
    __tablename__ = "forum_threads"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    title: Mapped[str] = mapped_column(String, nullable=False)
    date: Mapped[datetime.datetime] = mapped_column(DateTime, nullable=False)
    url: Mapped[str] = mapped_column(String, nullable=False)
    source: Mapped[str] = mapped_column(String, nullable=False)
    posts: Mapped[list["ForumPost"]] = relationship(back_populates="thread")
    content_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("json_content.id", ondelete="CASCADE"), nullable=False
    )
    content: Mapped[JSONContent] = relationship("JSONContent")

    def __repr__(self):
        return f"<ForumThread {self.title}>"


class ForumPost(Base):
    __tablename__ = "forum_posts"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    satoshi_id: Mapped[int] = mapped_column(Integer, unique=True, nullable=True)
    url: Mapped[str] = mapped_column(String, nullable=False)
    subject: Mapped[str] = mapped_column(String, nullable=False)
    poster_name: Mapped[str] = mapped_column(String, nullable=False)
    poster_url: Mapped[str] = mapped_column(String, nullable=True)
    date: Mapped[datetime.datetime] = mapped_column(DateTime, nullable=False)
    text: Mapped[str] = mapped_column(Text, nullable=False)
    disclaimer: Mapped[str] = mapped_column(String, nullable=True)
    nested_level: Mapped[int] = mapped_column(
        Integer,
        CheckConstraint("nested_level >= 0", name="nested_level"),
        nullable=False,
    )
    source_id: Mapped[str] = mapped_column(String, nullable=False)
    thread_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("forum_threads.id"), nullable=False
    )
    thread: Mapped[ForumThread] = relationship(back_populates="posts")
    quotes: Mapped[list["Quote"]] = relationship(
        back_populates="post", cascade="all, delete-orphan"
    )
    content_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("json_content.id", ondelete="CASCADE"), nullable=False
    )
    content: Mapped[JSONContent] = relationship("JSONContent")

    def __repr__(self):
        return f"<ForumPost {self.subject} - {self.source_id}>"
